### PR TITLE
[Merged by Bors] - feat: port Algebra.Homology.ComplexShape

### PR DESCRIPTION
**Pull request by joelriou**
(Not the most useful port at this stage. Just wanted to try some easy file.)

mathlib SHA : fd47bdf09e90f553519c712378e651975fe8c829

### CODE REVIEW

File: Mathlib/Algebra/Homology/ComplexShape.lean [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+when they exist, and return their input otherwise.
+
+This design aims to avoid certain problems arising from dependent type theory.
+In particular we never have to ensure morphisms `d i : X i ⟶ X (succ i)` compose as
+expected (which would often require rewriting by equations in the indexing type).
+Instead such identities become separate proof obligations when verifying that a
+complex we've constructed is of the desired shape.
+
+If `α` is an `AddRightCancelSemigroup`, then we define `up α : ComplexShape α`,
+the shape appropriate for cohomology,so `d : X i ⟶ X j` is nonzero only when `j = i + 1`,
```
Comment (semorrison) [inline Mathlib/Algebra/Homology/ComplexShape.lean:35]:
```suggestion
the shape appropriate for cohomology, so `d : X i ⟶ X j` is nonzero only when `j = i + 1`,
```

File: Mathlib/Algebra/Homology/ComplexShape.lean [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+This means that the shape consists of some union of lines, rays, intervals, and circles.
+
+Below we define `c.next` and `c.prev` which provide these related elements.
+-/
+@[ext]
+structure ComplexShape (ι : Type _) where
+  rel : ι → ι → Prop
+  next_eq : ∀ {i j j'}, rel i j → rel i j' → j = j'
+  prev_eq : ∀ {i i' j}, rel i j → rel i' j → i = i'
+#align complex_shape ComplexShape
```
Comment (semorrison) [inline Mathlib/Algebra/Homology/ComplexShape.lean:63]: If you decide no name change is necessary beyond what mathport suggested, it's allowed (and possibly preferred??) to delete the #align.

**Comment by semorrison** [comment]
Otherwise, if you could just satisfy the documentation linter (structure fields now need doc-strings), this will be great.

File: Mathlib/Algebra/Homology/ComplexShape.lean (hunk 2 above)
Comment (joelriou) [reply Mathlib/Algebra/Homology/ComplexShape.lean:67]: From the mathport output, the only name change I did was for the field `rel` (which mathport renamed `Rel`). Then, should I remove all the `#align` in this file?

Comment (digama0) [reply Mathlib/Algebra/Homology/ComplexShape.lean:67]: the field should be called `Rel`.

Comment (semorrison) [reply Mathlib/Algebra/Homology/ComplexShape.lean:67]: (Because its conclusion is a `Prop` or `Type`.)

**Comment by semorrison** [comment]
@joelriou, you'll also need to import this file in `Mathlib.lean`.

File: Mathlib/Algebra/Homology/ComplexShape.lean (hunk 2 above)
Comment (joelriou) [reply Mathlib/Algebra/Homology/ComplexShape.lean]: Thanks!

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**